cmd/node/node.go

```diff
@@ -115,20 +116,32 @@ func GetCommand() *cobra.Command {
 					log.NewWithLevel("", zap.NewAtomicLevelAt(zapcore.DebugLevel)),
 					events.EventHook())),
 			)
-			starter := func() error {
-				if err := app.Initialize(); err != nil {
+
+			run := func(ctx context.Context) error {
+				if err = app.Initialize(); err != nil {
 					return err
 				}
 				// This blocks until the context is finished or until an error is produced
-				if err := app.Start(); err != nil {
-					return err
+				err = app.Start(ctx)
+
+				ctx2, cancel2 := context.WithTimeout(context.Background(), 30*time.Second)
+				defer cancel2()
+				done := make(chan struct{}, 1)
+				go func() {
+					app.Cleanup(ctx2)
+					close(done)
+				}()
+				select {
+				case <-done:
+				case <-ctx2.Done():
+					log.With().Error("app failed to clean up in time")
 				}
-
-				return nil
+				return err
```

Comment:
I think this is more complicated than it needs to be. Can `app.Cleanup()` not react to the timeout itself?

```suggestion
				if err := app.Start(); err != nil {
					return err
				}
				
				ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
				defer cancel()
				return app.Cleanup(ctx)
```

Reply:
app.Cleanup will hang in this case. as the shutdown routines do not check the context when shutting down. 
refactoring that is more time than i want to spend on this.

Reply:
But then `app.Cleanup()` will also hang in the unchanged version of the code and `"app failed to clean up in time"` will be logged after 30 seconds before this function returns?

If this is the case can we add a `// FIXME` here and create an issue for it? `app.Cleanup` should definitely return when done cleaning up or when the `context` is canceled.

Reply:
my intention in the current code is to give up on cleaning up after 30 sec.
in this case, app will exit with the original app.Start() error. is that consistent with your reading?

i can create an issue for this yes.

Reply:
tho, even if we changed our codepath in the shutdown routine, the library we use may not support that. so i suspect the code here will remain unchanged even after the timeout ctx is plumbed through in our code.

Reply:
I'm just wondering why `app.Cleanup()` shouldn't be able to finish within 30 seconds. It unlocks via the `.LOCK` file and then tries to stop all our services. This would mean one of our services doesn't return when its `Stop()`/`Close()` method is called.

If we skip `app.Cleanup()` this could possibly lead to an inconsistent state.

I agree that this is out of scope for this issue but should be investigated.

Reply:
All `Close()` / `Stop()` calls are currently done sequentially, these could be done in parallel, e.g. with a `WaitGroup`; then every `Stop()` would have (about) the same timeout.

If one of the libraries we have doesn't have the option to pass in a `context` we should instead call the `Stop()` method of that component with something like this:

```golang
// within `app.stopServices(context.Context)`

var eg errgroup.Group
eg.Go(func() error {
	// component that respects context
	return component1.Stop(ctx)
})
eg.Go(func() error {
	errChan := make(chan error)
	go func() {
		err := component2.Stop()
		if err != nil {
			errChan <- fmt.Errorf("failed to stop component 2: %w", err)
		}
		close(errChan)
	}()
	
	select {
	case err := <-errCh:
		return err
	case <-ctx.Done():
		// possibly log here that component2.Stop() failed to return before context timed out
		return fmt.Errorf("failed to stop component 2: %w", ctx.Err())
	}
	return nil
})
// more components
return eg.Wait()
```

This way at least we can figure out which components cause the issues.

Reply:
filed https://github.com/spacemeshos/go-spacemesh/issues/3830 and added FIXME